xmppserver/src/main/java/org/jivesoftware/openfire/component/InternalComponentManager.java

```diff
@@ -708,29 +708,35 @@ public void joinedCluster( final byte[] nodeID )
     @Override
     public void leftCluster()
     {
-        // Upon leaving a cluster, the server uses its non-clustered/default ID again. Here, all clustered IDs are replaced with the new identity.
-        final NodeID defaultNodeID = XMPPServer.getInstance().getDefaultNodeID();
-        final NodeID nodeID = XMPPServer.getInstance().getNodeID();
-        if ( !defaultNodeID.equals( nodeID ) ) // In more recent versions of Openfire, the ID does not change.
-        {
-            CacheUtil.replaceValueInMultivaluedCache( componentCache, nodeID, defaultNodeID );
+        // The local cluster node left the cluster.
+        if (XMPPServer.getInstance().isShuttingDown()) {
+            // Do not put effort in restoring the correct state if we're shutting down anyway.
+            return;
         }
 
-        // The local cluster node left the cluster.
-        //
+        // Upon leaving a cluster, clustered caches are reset to their local equivalent (by the swap from the clustered
+        // cache implementation to the default cache implementation that's done in the implementation of
+        // org.jivesoftware.util.cache.CacheFactory.leftCluster). This means that they now hold no data (as a new cache
+        // has been created). Data that's available on the local node needs to be added again.
+        restoreCacheContent();
+
         // Determine what components that we are broadcasting were available only on other
         // cluster nodes than the local one. For these, unavailability events need
         // to be sent out (eventing will be limited to the local node, as we're no longer
         // part of a cluster).
-        final Map<Boolean, Map<JID, HashSet<NodeID>>> modified = CacheUtil.retainValueInMultiValuedCache( componentCache, defaultNodeID );
-
-        modified.get( false ).keySet().forEach( removedDomain -> {
-            Log.debug( "The local cluster node left the cluster. The component '{}' was living on one (or more) other cluster nodes, and is no longer available. Invoking the 'component unregistered' event.", removedDomain );
-            notifyComponentUnregistered( removedDomain );
-
-            // Also clean up the cache of cluster-wide service discovery results.
-            componentInfo.remove( removedDomain.toString() );
-        } );
+        final Iterator<Map.Entry<String, IQ>> iterator = componentInfo.entrySet().iterator();
+        while (iterator.hasNext()) {
+            // TODO: Iterating over the 'componentInfo' is not ideal, as there might be components for which there is no recorded data in 'componentInfo'. There currently is no other way to determine what components were active on the cluster though.
```

Comment:
It it worth raising a ticket so that this TODO isn't forgotten about ...

Reply:
Yes - and there's a similar TODO elsewhere that should also be captured. Will do after we're confident that this PR actually gets merged.

Reply:
https://issues.igniterealtime.org/browse/OF-2066
https://issues.igniterealtime.org/browse/OF-2067